### PR TITLE
Test Refactoring

### PR DESCRIPTION
**Pull request by jsamoocha**
This adds a new job to the (now renamed) `build-docs.yml` workflow file that runs flake8 and pytest (and as such replaces the existing nosetest framework by the more modern pytest). Please note that only the _unit_ and _integration_ tests are included in the CI build.

**TODO**:

- [x] Review
- [x] Update Contributing documentation 
- [x] Create separate actions for lint/test/doc, testing should also run on multiple OS

### CODE REVIEW

File: .github/workflows/build-test-docs.yml
Comment (yihong0618) [inline .github/workflows/build-test-docs.yml:47]: do we still support 2.7?

Comment (yihong0618) [reply .github/workflows/build-test-docs.yml:47]: https://github.com/stravalib/stravalib/issues/231

**Review by yihong0618 — COMMENTED** [review]
Is it better to separate `doc yml` and `test yml`

**Comment by yihong0618** [comment]
I think I will do `drop python2` work later we can save some codes.

File: .github/workflows/build-test-docs.yml
Comment (jsamoocha) [reply .github/workflows/build-test-docs.yml:47]: > Is it better to separate `doc yml` and `test yml`

Sorry for my confusion: is this an open question or suggestion for improvement? I don't consider myself an expert on Github Actions, so any advice is welcome.

> do we still support 2.7?

Good question, TBH I just copy-pasted the versions mentioned in the starter workflow. We may also need to add the recent 3.11?

Comment (yihong0618) [reply .github/workflows/build-test-docs.yml:47]: no need to add 3.11 for now I think.

Comment (yihong0618) [reply .github/workflows/build-test-docs.yml:47]: > Sorry for my confusion: is this an open question or suggestion for improvement? I don't consider myself an expert on Github Actions, so any advice is welcome.

its an open question.~

**Comment by jsamoocha** [comment]
I added minimal working examples of testing client endpoints with a mocked Strava API.

The mocked Strava API can be used as a pytest fixture and is implemented as a thin wrapper around the `RequestsMock` type from the `responses` package. Instead of specifying the entire expected JSON response body for every request, the published example responses by Strava (see the swagger.json file in RESOURCES) are used. They can be tweaked by providing a `response_update` argument to the mock methods.

TODO: dynamically choosing a response from the swagger.json based on the status code in case of non-200 responses.

@yihong0618 @lwasser @hozn, please let me know if this approach works for you. If so, we can continue migrating the other tests.

**Comment by yihong0618** [comment]
> I added minimal working examples of testing client endpoints with a mocked Strava API.
> 
> The mocked Strava API can be used as a pytest fixture and is implemented as a thin wrapper around the `RequestsMock` type from the `responses` package. Instead of specifying the entire expected JSON response body for every request, the published example responses by Strava (see the swagger.json file in RESOURCES) are used. They can be tweaked by providing a `response_update` argument to the mock methods.
> 
> TODO: dynamically choosing a response from the swagger.json based on the status code in case of non-200 responses.
> 
> @yihong0618 @lwasser @hozn, please let me know if this approach works for you. If so, we can continue migrating the other tests.

Looks good to me.

**Comment by lwasser** [comment]
You rock! I will look today.

**Comment by jsamoocha** [comment]
> Looks good to me. can this `PR` be reviewed now or still WIP

Yes, please review it when you have time. The module `tests.integration.test_client` has some examples of how to use the mock for different purposes, which may be a good starting point.

File: .github/workflows/build-test-docs.yml
Comment (yihong0618) [inline .github/workflows/build-test-docs.yml:9]: when pull request only master branch is OK

File: stravalib/tests/integration/strava_api_stub.py
Comment (yihong0618) [inline stravalib/tests/integration/strava_api_stub.py:48]: FYI, if we use type hint here we can not run these tests in python2.
But its fine, since we will drop python2 later

**Review by yihong0618 — APPROVED** [review]
LGTM, amazing work!

File: .github/workflows/build-test-docs.yml
Comment (lwasser) [reply .github/workflows/build-test-docs.yml:47]: hi! i think dropping python 2.x makes sense and we should do it. Most packages are doing this as it's end of life. i think generally supporting the latest several versions of python makes sense. we can always add 3.11 later as well. 

i'm going to look at the full build here in just a few but we should also setup a matrix to test on windows, mac and linux -

File: .github/workflows/build-test-docs.yml
Comment (lwasser) [inline .github/workflows/build-test-docs.yml:9]:
```suggestion
on:
  pull_request:
  push:
    branches:
      - master
```
This will then build on merge to master and on pr but won't try to do both when someone submits a pr

File: .github/workflows/build-test-docs.yml
Comment (lwasser) [reply .github/workflows/build-test-docs.yml:47]: so i think we should add a matrix  build. in this example we'd break the build into a docs only build that runs on linux or whatever os we want. Then a second that uses the matrix to run tests across versions of python and operating systems. 

Here is an example of what the matrix build would look like. i'm happy to help with this

https://github.com/earthlab/abc-classroom/blob/main/.github/workflows/run-tests-mac-win-linux.yml#L14
But we don't need in this example the various environments, we just need the os matrix part 
```
  setup-build:
    name: Ex1 (${{ matrix.python-version }}, ${{ matrix.os }})
    runs-on: ${{ matrix.os }}
    strategy:
      fail-fast: false
      matrix:
        os: ["ubuntu-latest", "macos-latest", "windows-latest"]
        python-version: [3.7, 3.8, 3.9]

    steps:
    - uses: actions/checkout@v2
    - name: Set up Python ${{ matrix.python-version }}
      uses: actions/setup-python@v2
      with:
        python-version: ${{ matrix.python-version }}
```

File: .github/workflows/build-test-docs.yml
Comment (lwasser) [inline .github/workflows/build-test-docs.yml:76]: so psyched you're using pytest!! yay!

File: .github/workflows/build-test-docs.yml
Comment (lwasser) [inline .github/workflows/build-test-docs.yml:70]: so we wouldn't need to lint across os' either. i can ask some friends tomorrow about how they handle linting vs doc build vs tests where you really want a matrix. but my gut is telling me a build for tests and a build for docs and linting OR conditional statements here that only look at docs on one element of the workflow matrix ie 

if python 3.8 and mac os, build docs, lint run tests
else 
only run tests

File: README.md
Comment (lwasser) [inline README.md:78]: is this replaced by the mock element? i haven't dug into the actual tests yet :)

File: stravalib/protocol.py
Comment (lwasser) [inline stravalib/protocol.py:166]: is this a helper? i'm curious is the rename just for clarity? i'm fine either way. i'm used to [helper  / non public methods using an underscore to follow PEP8 ](https://peps.python.org/pep-0008/#method-names-and-instance-variables)

File: stravalib/tests/integration/strava_api_stub.py
Comment (lwasser) [reply stravalib/tests/integration/strava_api_stub.py:48]: if y'all are ok with dropping python 2.x i really think we should!

**Review by lwasser — COMMENTED** [review]
This looks really good to me. I am traveling for the next few days and wont be able to run this locally. I defer to you as the expert on how the test suite is designed but would love to better understand how the mock works. I'm wondering if you're using that swagger json response as the expected structure of the return from the actual API?

Kind of like we might do with a cassette but that might be better and simpler? 

In terms of changes - i do think we should run the test suite on a matrix. And then let's just have separate builds for 
tests
docs 
linting? or combine docs and linting?

 a separate build might be easier to look at from a maintenance perspective because with the matrix will will then have python 3.9, windows, 3.9 mac etc etc...

Here is a build from one of my favorite people who is really creating nice clean docs and such in their community. They have a build for  tests, docs, style (linting)

https://github.com/fatiando/boule/tree/main/.github/workflows 

let me know if i can help here!

**Comment by yihong0618** [comment]
> if y'all are ok with dropping python 2.x i really think we should!
will do this after this `PR` merge

File: stravalib/tests/integration/strava_api_stub.py
Comment (jsamoocha) [reply stravalib/tests/integration/strava_api_stub.py:48]: I guess we agree that we'll not be supporting Python 2 any longer, so I'll keep the type hints.

File: .github/workflows/build-test-docs.yml
Comment (jsamoocha) [reply .github/workflows/build-test-docs.yml:76]: Supposedly, it runs the legacy (unittest) tests without any changes. So why not? ¯\_(ツ)_/¯

File: README.md
Comment (jsamoocha) [reply README.md:78]: I still need to modify the documentation in a.o. README.md, based on this review.

File: stravalib/protocol.py
Comment (jsamoocha) [reply stravalib/protocol.py:166]: I decided to reuse this method outside of the `ApiV3` class to help set up the mocks, so made this method public. Alternatively, we could also refactor by completely moving the server address, API base, and resolve_url function to some generic util module or package __init__ module. Happy to make that change if you think the latter is necessary.

File: .github/workflows/build-test-docs.yml
Comment (jsamoocha) [reply .github/workflows/build-test-docs.yml:47]: Ok, I'll create separate actions for test/lint/doc, where test is using a matrix as you mentioned above.

File: stravalib/protocol.py
Comment (lwasser) [reply stravalib/protocol.py:166]: @jsamoocha got it. this makes sense. I will default to your expertise here in setting this up. What would be the benefit of a complete refactor?

thank you again for this work here. it's awesome to have you here helping maintain stravalib.

File: stravalib/tests/integration/conftest.py
Comment (hozn) [reply stravalib/tests/integration/conftest.py:8]: @lwasser , I'll jump in and provide a little context around this.  What @jsamoocha is doing -- and honestly, this is brilliant -- is using Strava's published Swagger/OpenAPI JSON (https://developers.strava.com/swagger/swagger.json) to extract all the known API paths and the examples (which are in that document) that Strava provides as a sort of documentation.  To address your question more specifically, the mock here serves to intercept and replace what would be the real HTTP requests out to the Strava API (which would require an application ID / secret / athletes / etc.) with the responses that were provided by Strava as examples in their Swagger.

I particularly like this approach as:
1. As mentioned above, this removes the need for having to set up any real infrastructure (accounts / apps / etc.) to run the tests
2. The example responses are pulled down with every test run.  Since these will be kept up-to-date by Strava we'll know that if stravalib starts breaking that something in the API is no longer as expected.

Comment (jsamoocha) [reply stravalib/tests/integration/conftest.py:8]: Thanks for the explanation, @hozn, while I was busy adding some documentation to the code!

@lwasser, to address your question regarding passthru: because the fetching of the swagger.json file from Strava is happening within the context of the mocked API, that request would've been mocked as well. So we need to use `add_passthru()` to tell the mocker not to interfere with that request.

Comment (lwasser) [reply stravalib/tests/integration/conftest.py:8]: awesome. @jsamoocha thank you for all of this work and for the explanation about how this is working. i'll defer to any issues @hozn sees in the PR which it sounds like they really appreciate your approach !! so please don't let me slow down the process of merging this. it all looks good to me !

File: .github/workflows/build-test-docs.yml
Comment (jsamoocha) [reply .github/workflows/build-test-docs.yml]: Yes, good suggestion! It is included in the 3 separate workflow files now.

**Pull request merged**
